x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java

```diff
@@ -148,6 +152,54 @@ public Automaton allowedActionsMatcher(String index) {
         return automatonList.isEmpty() ? Automatons.EMPTY : Automatons.unionAndMinimize(automatonList);
     }
 
+    /**
+     * Determines if this {@link IndicesPermission} is a subset of other indices
+     * permission. This iteratively determines if all of the
+     * {@link IndicesPermission.Group} for this IndicesPermission is a subset of
+     * some group in the given indices permission.
+     *
+     * @param other indices permission
+     * @return in case it is not a subset then returns
+     * {@link SubsetResult.Result#NO} and if it is clearly a subset will return
+     * {@link SubsetResult.Result#YES}. It will return
+     * {@link SubsetResult.Result#MAYBE} when the role is a subset in every
+     * other aspect except DLS queries since we cannot determine if a query
+     * returns a subset of documents
+     */
+    public SubsetResult isSubsetOf(final IndicesPermission other) {
+        if (this.groups() == null || this.groups().length == 0) {
+            return SubsetResult.isASubset();
+        }
+        Set<Set<String>> combineDLSQueriesFromIndexPrivilegeMatchingTheseNames = new HashSet<>();
+        for (Group thisGroup : this.groups()) {
+            boolean granted = false;
+            Set<Set<String>> maybeGroupIndices = Collections.emptySet();
+            if (other.groups() != null) {
+                for (Group otherGroup : other.groups()) {
+                    final SubsetResult result = thisGroup.isSubsetOf(otherGroup);
+                    if (result.result() == Result.YES) {
+                        granted = true;
+                        maybeGroupIndices.clear();
```

Comment:
I'm tired, so I'm not quite following this. Why do we clear this set when we find a single `YES` ?

Depite being declared outside the `for` loop, it seems to only ever hold the value from the last `otherGroup`